### PR TITLE
Stackdriver deep link support for multiple logged in accounts

### PR DESCRIPTION
**Pull request by papagian**
**What this PR does / why we need it**:
Adds a new input in the metric query editor where the user can optionally provide the google account that the deep link should target.
If such an account is provided and is valid, `authuser` query parameter will be included in the deep link.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
The present PR targets #25858

### CODE REVIEW

File: docs/sources/features/datasources/stackdriver.md [hunk 1]
```diff
@@ -333,4 +333,7 @@ Only available in Grafana v7.1+.
 > **Note:** This feature is only available for Metric queries.
 
 Click on a time series in the panel to see a context menu with a link to View in Monitoring Metrics explorer. Clicking that link opens the Metrics explorer in the Monitoring and runs the query from the Grafana panel there.
-If you're not currently logged in to the Google console, then the link opens the login page. The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
+
+The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
+
+If you're not currently logged in to the Google console, then the link opens the login page. If you are logged in in multiple google accounts you’ll directed to the first account you signed into. This behaviour can be ovewritten if you fill the Google Account input in the metric query editor.
```
Comment (aknuds1) [inline docs/sources/features/datasources/stackdriver.md:339]: Spotted some typos and some potential improvements, I also think it should be overridden and not overwritten?
```suggestion
If you're not currently logged in to the Google console, the link opens the login page. If you are logged in to multiple Google accounts you’ll be directed to the first account you signed in to. This behaviour can be overridden if you fill the Google Account input into the metric query editor.
```

File: pkg/tsdb/stackdriver/stackdriver.go [hunk 2]
```diff
@@ -128,6 +137,13 @@ func (query *stackdriverQuery) buildDeepLink() string {
 
 	q := u.Query()
 	q.Set("project", query.ProjectName)
+	if query.GoogleAccount != "" {
+		if !isValidEmail(query.GoogleAccount) {
+			slog.Debug("Invalid google account is not added to the deep link", "account", query.GoogleAccount)
```
Comment (aknuds1) [inline pkg/tsdb/stackdriver/stackdriver.go:142]:
```suggestion
			slog.Debug("Invalid Google account is not added to the deep link", "account", query.GoogleAccount)
```

File: pkg/tsdb/stackdriver/stackdriver.go [hunk 3]
```diff
@@ -116,6 +116,15 @@ func (query *stackdriverQuery) isSLO() bool {
 	return query.Slo != ""
 }
 
+func isValidEmail(email string) bool {
+	if email == "" {
+		return true
+	}
+
+	re := regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
```
Comment (aknuds1) [inline pkg/tsdb/stackdriver/stackdriver.go:124]: Is this impressive regex copied from StackOverflow? I recall seeing one there :)

File: pkg/tsdb/stackdriver/stackdriver_test.go [hunk 4]
```diff
@@ -68,10 +68,51 @@ func TestStackdriver(t *testing.T) {
 						"perSeriesAligner": "ALIGN_MEAN",
 						"filter":           "metric.type=\"a/metric/type\"",
 					}
-					verifyDeepLink(dl, expectedTimeSelection, expectedTimeSeriesFilter)
+					params := verifyDeepLink(dl, expectedTimeSelection, expectedTimeSeriesFilter)
+					So(params.Get("authuser"), ShouldBeEmpty)
 				})
 			})
 
+			Convey("and query has google account", func() {
```
Comment (aknuds1) [inline pkg/tsdb/stackdriver/stackdriver_test.go:76]:
```suggestion
			Convey("and query has Google account", func() {
```

File: pkg/tsdb/stackdriver/stackdriver_test.go [hunk 5]
```diff
@@ -68,10 +68,51 @@ func TestStackdriver(t *testing.T) {
...
+				}
+				expectedTimeSeriesFilter := map[string]interface{}{
+					"perSeriesAligner": "ALIGN_MEAN",
+					"filter":           "metric.type=\"a/metric/type\"",
+				}
+				params := verifyDeepLink(dl, expectedTimeSelection, expectedTimeSeriesFilter)
+				So(params.Get("authuser"), ShouldEqual, "foo@google.com")
+			})
+
+			Convey("and query has invalid google account", func() {
```
Comment (aknuds1) [inline pkg/tsdb/stackdriver/stackdriver_test.go:96]:
```suggestion
			Convey("and query has invalid Google account", func() {
```

File: public/app/plugins/datasource/stackdriver/components/GoogleAccount.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  value: string;
+}
+
+export const GoogleAccount: FunctionComponent<Props> = ({ value = '', onChange }) => {
+  const [googleAccount, setGoogleAccount] = useState(value);
+
+  return (
+    <QueryInlineField
+      label="Google Account"
+      tooltip="The google account passed to the deep link (useful if the user is signed in to multiple accounts)"
```
Comment (aknuds1) [inline public/app/plugins/datasource/stackdriver/components/GoogleAccount.tsx:15]:
```suggestion
      tooltip="The Google account passed to the deep link (useful if the user is signed in to multiple accounts)"
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Spotted some typos and potential language improvements.

File: pkg/tsdb/stackdriver/stackdriver.go (hunk 3 above)
Comment (papagian) [reply pkg/tsdb/stackdriver/stackdriver.go:124]: I no more recall from where I have copied, but I didn't compiled it myself for sure.

**Review by aknuds1 — APPROVED** [review]

**Comment by daniellee** [comment]
@papagian @marefr this solution works but will only work for one person. To solve that, we could have a list of email accounts but wondering if the query editor is the right place for this setting?

**Review by marefr — COMMENTED** [review]
Interesting approach, but I'm skeptic as well. Maybe skip this and ship the first version and wait for feedback.

**Comment by marefr** [comment]
Best thing would be if Google supported forcing to prompt for which account to use

File: docs/sources/features/datasources/stackdriver.md [hunk 7]
```diff
@@ -333,4 +333,7 @@ Only available in Grafana v7.1+.
 > **Note:** This feature is only available for Metric queries.
 
 Click on a time series in the panel to see a context menu with a link to View in Monitoring Metrics explorer. Clicking that link opens the Metrics explorer in the Monitoring and runs the query from the Grafana panel there.
-If you're not currently logged in to the Google console, then the link opens the login page. The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
+
+The provided link is valid for any account, but it only displays the query if your account has access to the GCP project specified in the query.
+
+If you're not currently logged in to the Google console, the link opens the login page. If you are logged in to multiple Google accounts you’ll be directed to the first account you signed in to. This behaviour can be overridden if you fill the Google Account input into the metric query editor.
```
Comment (oddlittlebird) [inline docs/sources/features/datasources/stackdriver.md:339]:
```suggestion
If you're not currently logged in to the Google console, then the link opens the login page. If you are logged in to multiple Google accounts, then you’ll be directed to the first account you signed in to. You can override this behavior can if you fill the Google Account input into the metric query editor.
```

**Comment by papagian** [comment]
Closing this in favour of the solution described [here](https://github.com/grafana/grafana/pull/25858#issuecomment-651675956).

**Pull request closed without merging**